envoy-control-core/src/main/kotlin/pl/allegro/tech/servicemesh/envoycontrol/snapshot/resource/endpoints/EnvoyEndpointsFactory.kt

```diff
@@ -47,6 +48,79 @@ class EnvoyEndpointsFactory(
             }
     }
 
+    fun filterEndpoints(
+        clusterLoadAssignment: ClusterLoadAssignment,
+        routingPolicy: RoutingPolicy
+    ): ClusterLoadAssignment {
+        if (!routingPolicy.autoServiceTag || !properties.routing.serviceTags.enabled) {
+            return clusterLoadAssignment
+        }
+
+        val filteredLoadAssignment = routingPolicy.serviceTagPreference.firstNotNullOfOrNull { serviceTag ->
```

Comment:
i think it can be simplified - both `containsEndpointWithServiceTag` and `createFilteredLoadAssignment` iterate through same lists and filter it using `metadataContainsServiceTag`

Reply:
and then we could get rid of this `ContainsResult`

Reply:
True, but the idea is to not create `ClusterLoadAssignement` unnecessary. 
`containsEndpointWithServiceTag` and `createFilteredLoadAssignment` indeed iterate over the same thing - but `containsEndpointWithServiceTag` only iterate and `createFilteredLoadAssignment` also create new object.

That's why I intentionally splitted this into 2 iterations. Let's think for example about a environment where 100% envoys don't use this feature - if we check and build new `ClusterLoadAssignement` we will waste tons of memory, cpu, GC, etc.

Reply:
it looks to me like it can still be simplified, keeping the acutal performance

Reply:
I tried and wirtten something like it
```
    private fun filterEndpoints(loadAssignment: ClusterLoadAssignment, tag: String): ClusterLoadAssignment {
        var anyEndpointMatched = false
        var allEndpointMatched = true
        val filteredEndpoints = loadAssignment.endpointsList.map { localityLbEndpint ->
            val (matchedEndpoints, unmatchedEndpoints) = localityLbEndpint.lbEndpointsList.partition { metadataContainsServiceTag(it.metadata, tag) }
            when  {
                matchedEndpoints.isNotEmpty() && unmatchedEndpoints.isNotEmpty() -> { // SOME
                    anyEndpointMatched = true
                    allEndpointMatched = false
                    localityLbEndpint.toBuilder()
                        .clearLbEndpoints()
                        .addAllLbEndpoints(matchedEndpoints)
                        .build()
                }
                matchedEndpoints.isNotEmpty()  && unmatchedEndpoints.isEmpty() -> { // ALL
                    localityLbEndpint
                }
                else -> { // NONE
                    allEndpointMatched = false
                    null
                }
            }
        }
        return when {
            allEndpointMatched -> loadAssignment // ALL
            anyEndpointMatched -> loadAssignment.toBuilder()  // SOME
                .clearEndpoints()
                .addAllEndpoints(filteredEndpoints)
                .build()
            else -> createEmptyLoadAssignment(loadAssignment) // NONE
        }
    }
```

Reply:
I agree, it's better. I implemented it, even managed to simplify it more. 
But the tests show there is a small change in the results - it filters out empty localities entirely.
You can see it there: https://github.com/allegro/envoy-control/pull/360/files#diff-e10d991433468a07369772731e0126754d2fa8b286d796a00d8144824cd4da01L130

For now I created it as another PR to this PR, because I don't know if such change does matter, and if it does, is it for better or worse. Let me know what do you think

Reply:
i think it is acceptable - sending clusterLoadAssigments without endpoints or not sending, doesn't change final endpoints list